File: uprank-app/types/job.ts
// UUID type definition for simplicity
type UUID = string;

// Schema platform type definition
type Platform = string;


// JobEdges type definition
interface JobEdges {
    upworkjob?: UpworkJob;
    loadedTypes: boolean[];
}

// UpworkJob type definition
export interface UpworkJob {
    id?: string;
    title?: string;
    created_at?: Date;
    updated_at?: Date;
    embedded_at?: Date;
    ranked_at?: Date;
    location?: string;
    description?: string;
    skills?: string[];
    experience_level?: string;
    hourly?: boolean;
    fixed?: boolean;
    hourly_rate?: number[];
    fixed_rate?: number;
    average_uprank_score?: number;
    max_uprank_score?: number;
    min_uprank_score?: number;
    edges: UpworkJobEdges;
    job_upworkjob?: UUID;
}

// UpworkJobEdges type definition
interface UpworkJobEdges {
    // Define any necessary fields here
}

// Job type definition
export interface Job {
    id?: UUID;
    origin_platform?: Platform;
    edges: JobEdges;
    user_job?: string;
}

File: uprank-backend/main-backend/ent/schema/job.go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/dialect"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
)

type Job struct {
	ent.Schema
}

func (Job) Fields() []ent.Field {
	return []ent.Field{
		field.String("id").NotEmpty().
			Unique().
			Immutable(),
		field.String("title").NotEmpty(),
		field.Time("created_at").
			Default(time.Now),
		field.String("location").Optional(),
		field.String("description").NotEmpty(),
		field.JSON("skills", []string{}).Optional(),
		field.String("experience_level").Optional(),
		field.Bool("hourly"),
		field.Bool("fixed"),
		field.JSON("hourly_rate", []float32{}).Optional(), // hourly rate can be an array of floats with length 1 or 2 (for a range of hourly rates)
		field.Float("fixed_rate").
			SchemaType(map[string]string{
				dialect.Postgres: "numeric",
			}).Optional(),
		field.Float("average_uprank_score").Optional(),
		field.Float("max_uprank_score").Optional(),
		field.Float("min_uprank_score").Optional(),
	}
}

// Edges of the Job.
func (Job) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("user", User.Type).
			Ref("jobs").
			Unique().Required(),
		edge.To("freelancers", Freelancer.Type),
	}
}
